### ISSUE TITLE
On successful bind user details not saved in a local Django User model and user not logged in

### ISSUE THREAD

**Issue by uostim** [etianen/django-python3-ldap#65]
Hi,

I've just created a fresh Django project with all default settings (default user model etc.) to test this. 

I've installed django-python3-ldap as described, added 'django_python3_ldap' my INSTALLED_APPS and set AUTHENTICATION_BACKENDS = ['django_python3_ldap.auth.LDAPBackend'].

The rest of the settings I've overridden are:

LDAP_AUTH_URL = "[my server URL]"
LDAP_AUTH_SEARCH_BASE = "dc=[1st DC],dc=[2nd DC],dc=[3rd DC],dc=[4th DC]"
LDAP_AUTH_FORMAT_USERNAME = "django_python3_ldap.utils.format_username_active_directory"

LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "handlers": {
        "console": {
            "class": "logging.StreamHandler",
        },
    },
    "loggers": {
        "django_python3_ldap": {
            "handlers": ["console"],
            "level": "DEBUG",
        },
    },
}

I used a python shell to test binding as follows:

>>from ldap3 import Server, Connection
>>s = Server('[Server URL]', get_info=ALL)
>>c = Connection(s, user='[username]', password='[right password]')
>>if not c.bind():
>>    print("Error!", c.result)
>>else:
>>    print("It Worked!", c.result)

>> It Worked! {'description': 'success', 'referrals': None, etc....}

>>c = Connection(s, user='[username]', password='[wrong password]')
>>if not c.bind():
>>    print("Error!", c.result)
>>else:
>>    print("It Worked!", c.result)

>> Error! {'description': 'invalidCredentials', 'referrals': None, etc....}

So now I'm trying to use this to log in to my django site. At first I was just trying to log into the admin site using this, but now I've built a proper login form in the front-end site just to check that this all works and it wasn't the admin login page doing something funky.

When I attempt to log in using an incorrect User & Pass, in my console I see:

[17/Mar/2017 12:56:27] "POST /accounts/login/ HTTP/1.1" 200 1135
LDAP login failed: automatic bind not successful - invalidCredentials

but when I use the correct User & Pass I only get:
[17/Mar/2017 12:56:27] "POST /accounts/login/ HTTP/1.1" 200 1135

The login page still says "Incorrect username and / or password! Please double-check and try again." and doesn't forward me on to the correct place, and there are no new user entries in my auth_user table in my DB.

Any idea's where I'm going wrong?  It appears I have enough settings right for it to recognise bad credentials, but not enough right for it to do anything useful on a successful bind!

Any help appreciated!

Many thanks,
-Tim

**Comment by etianen** [comment]
Authentication is in two parts.

1. Attempt to bind. This works.
2. Look up the user's details from the LDAP server, and use them to create a user model.

If it can't find the user details in the LDAP server, then authentication will still fail, even if the LDAP bind works.

Your LDAP_AUTH_SEARCH_BASE, or LDAP_AUTH_OBJECT_CLASS may be wrong. I'm a bit suspicious of your LDAP_AUTH_SEARCH_BASE, since the seach base is supposed to describe a path to the location of your users in the LDAP tree, but yours appears to be a set of *or* possibilities. If your users exist in multiple locations in the tree, select the nearest common ancestor and specify that as the search base.

**Comment by etianen** [comment]
Alternatively, the authenticated user might not have permission to query the LDAP database.

Upgrade to the master branch of django-python3-ldap, then try hardcoding `LDAP_AUTH_CONNECTION_USERNAME` and `LDAP_AUTH_CONNECTION_PASSWORD` in your settings, then running `ldap_sync_users` to chek your details.

**Comment by uostim** [comment]
Thanks for the suggestions etianen! I've reviewed my settings.py and our AD infrastructure with AD Explorer and I think I've found the problem, I think it was binding OK but not actually creating my user model entry because I also needed to overwrite the LDAP_AUTH_OBJECT_CLASS to be "organizationalPerson"!  I hadn't thought to check that our AD Object Class was anything other than the default... AD/LDAP intricacies are still something of a magical mystery to me!

Seems to all be working OK now though, I get user model entries created on successful bind. 

Thanks for your help and the time you've spend building and supporting this very useful package!
